src/sparse/csmat.rs

```diff
@@ -2826,6 +2826,25 @@ mod test {
         assert_eq!(diag, expected);
     }
 
+    #[test]
+    fn diag_using_shrink_to_fit() {
+        // | 1 0 0 3 1 |
+        // | 0 0 0 0 0 |
+        // | 0 0 0 1 0 |
+        // | 3 0 1 0 0 |
+        // | 1 0 0 0 0 |
+        let mat = CsMat::new_csc(
+            (5, 5),
+            vec![0, 3, 3, 4, 6, 7],
+            vec![0, 3, 4, 3, 0, 2, 0],
+            vec![1, 3, 1, 1, 3, 1, 1],
+        );
+
+        let diag = mat.diag();
+        let expected = CsVec::new(5, vec![0], vec![1]);
+        assert_eq!(diag, expected);
```

Comment:
This test does not actually test the capacity of the vector. I'm afraid there's currently no way to get the capacity of the vectors backing a `CsVec`.

You can add it if you want, it should be something like

```rust
    /// The capacity of the vectors backing the storage
    pub fn capacity(&self) -> usize {
        // All methods guarantee that `self.indices` and `self.data` have
        // the same capacity
        self.data.capacity()
    }
```

Reply:
But you don't have to, we can trust the standard library to correctly shrink to fit.

Reply:
That gives me an error. It seems like `capacity` can only be used on vectors and not slices?
<img width="757" alt="Screen Shot 2020-12-18 at 11 15 16 PM" src="https://user-images.githubusercontent.com/6342654/102666317-e5b1f400-4186-11eb-9315-fab26c45ff3b.png">

Reply:
The method needs to be on `CsVecI`, the owned version of `CsVecBase`.